File: beta_testing_platform_code/beta_testing_platform/src/routes/feedback.py
from flask import Blueprint, jsonify, request
from src.models.feedback import Feedback, db
import os
from datetime import datetime

# Try to import gspread, but make it optional for deployment
try:
    import gspread
    GSPREAD_AVAILABLE = True
except ImportError:
    GSPREAD_AVAILABLE = False

feedback_bp = Blueprint('feedback', __name__)

# Google Sheets configuration
SPREADSHEET_NAME = 'Beta Tester Feedback'
SERVICE_ACCOUNT_KEY_FILE = os.path.join(os.path.dirname(__file__), '..', 'config', 'service_account.json')

def append_to_google_sheet(feedback_data):
    """Append feedback data to Google Sheets"""
    if not GSPREAD_AVAILABLE:
        print("gspread not available - skipping Google Sheets sync")
        return False
        
    try:
        # Check if service account file exists
        if not os.path.exists(SERVICE_ACCOUNT_KEY_FILE):
            print(f"Service account file not found at {SERVICE_ACCOUNT_KEY_FILE}")
            return False
            
        # Authenticate with Google Sheets
        gc = gspread.service_account(filename=SERVICE_ACCOUNT_KEY_FILE)
        
        # Open the spreadsheet
        spreadsheet = gc.open(SPREADSHEET_NAME)
        worksheet = spreadsheet.sheet1
        
        # Prepare row data
        row_data = [
            feedback_data['timestamp'],
            feedback_data['tester_name'],
            feedback_data['submission_type'],
            feedback_data['title'],
            feedback_data['description'],
            feedback_data.get('severity', ''),
            feedback_data.get('status', 'New')
        ]
        
        # Append the row
        worksheet.append_row(row_data)
        return True
        
    except Exception as e:
        print(f"Error appending to Google Sheets: {e}")
        return False

@feedback_bp.route('/feedback', methods=['POST'])
def submit_feedback():
    """Submit new feedback/bug report/progress update"""
    try:
        data = request.json
        
        # Validate required fields
        required_fields = ['tester_name', 'submission_type', 'title', 'description']
        for field in required_fields:
            if field not in data or not data[field]:
                return jsonify({'error': f'Missing required field: {field}'}), 400
        
        # Create feedback record
        feedback = Feedback(
            tester_name=data['tester_name'],
            submission_type=data['submission_type'],
            title=data['title'],
            description=data['description'],
            severity=data.get('severity'),
            status=data.get('status', 'New')
        )
        
        # Save to database
        db.session.add(feedback)
        db.session.commit()
        
        # Prepare data for Google Sheets
        feedback_data = {
            'timestamp': feedback.timestamp.strftime('%Y-%m-%d %H:%M:%S'),
            'tester_name': feedback.tester_name,
            'submission_type': feedback.submission_type,
            'title': feedback.title,
            'description': feedback.description,
            'severity': feedback.severity or '',
            'status': feedback.status
        }
        
        # Append to Google Sheets
        sheets_success = append_to_google_sheet(feedback_data)
        
        response_data = feedback.to_dict()
        response_data['sheets_synced'] = sheets_success
        
        return jsonify(response_data), 201
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@feedback_bp.route('/feedback', methods=['GET'])
def get_feedback():
    """Get all feedback entries"""
    try:
        feedback_entries = Feedback.query.order_by(Feedback.timestamp.desc()).all()
        return jsonify([entry.to_dict() for entry in feedback_entries])
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@feedback_bp.route('/feedback/<int:feedback_id>', methods=['GET'])
def get_feedback_by_id(feedback_id):
    """Get specific feedback entry"""
    try:
        feedback = Feedback.query.get_or_404(feedback_id)
        return jsonify(feedback.to_dict())
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: netlify/functions/feedback.js
const { google } = require('googleapis');

// Google Sheets configuration
const SPREADSHEET_NAME = 'Beta Tester Feedback';

// Helper function to get Google Sheets client
async function getSheetsClient() {
  try {
    const serviceAccountJson = process.env.GOOGLE_SERVICE_ACCOUNT;
    if (!serviceAccountJson) {
      throw new Error('GOOGLE_SERVICE_ACCOUNT environment variable not set');
    }

    const serviceAccount = JSON.parse(serviceAccountJson);
    const auth = new google.auth.GoogleAuth({
      credentials: serviceAccount,
      scopes: ['https://www.googleapis.com/auth/spreadsheets']
    });

    return google.sheets({ version: 'v4', auth });
  } catch (error) {
    console.error('Error setting up Google Sheets client:', error);
    throw error;
  }
}

// Helper function to append to Google Sheets
async function appendToGoogleSheet(feedbackData) {
  try {
    const sheets = await getSheetsClient();
    
    // Prepare row data
    const rowData = [
      feedbackData.timestamp,
      feedbackData.tester_name,
      feedbackData.submission_type,
      feedbackData.title,
      feedbackData.description,
      feedbackData.severity || '',
      feedbackData.status || 'New'
    ];

    // Get the spreadsheet ID from environment variable or use a default
    const spreadsheetId = process.env.GOOGLE_SHEET_ID || '1BxiMVs0XRA5nFMdKvBdBZjgmUUqptlbs74OgvE2upms'; // Replace with your actual sheet ID
    
    // Append the row to the spreadsheet
    const response = await sheets.spreadsheets.values.append({
      spreadsheetId: spreadsheetId,
      range: 'Sheet1!A:G',
      valueInputOption: 'RAW',
      resource: {
        values: [rowData]
      }
    });

    return response.data;
  } catch (error) {
    console.error('Error appending to Google Sheets:', error);
    throw error;
  }
}

exports.handler = async (event, context) => {
  // Set CORS headers
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Content-Type': 'application/json'
  };

  // Handle preflight requests
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: ''
    };
  }

  try {
    if (event.httpMethod === 'GET') {
      // Get all feedback (for now, return empty array since we don't have a database)
      return {
        statusCode: 200,
        headers,
        body: JSON.stringify([])
      };
    }

    if (event.httpMethod === 'POST') {
      const feedbackData = JSON.parse(event.body);
      
      // Validate required fields
      const requiredFields = ['tester_name', 'submission_type', 'title', 'description'];
      for (const field of requiredFields) {
        if (!feedbackData[field]) {
          return {
            statusCode: 400,
            headers,
            body: JSON.stringify({ error: `Missing required field: ${field}` })
          };
        }
      }

      // Add timestamp and status
      const processedData = {
        ...feedbackData,
        timestamp: new Date().toISOString(),
        status: 'New'
      };

      // Try to append to Google Sheets
      let sheetsSuccess = false;
      try {
        await appendToGoogleSheet(processedData);
        sheetsSuccess = true;
      } catch (error) {
        console.error('Google Sheets sync failed:', error);
        // Continue without failing the request
      }

      // Return success response
      return {
        statusCode: 201,
        headers,
        body: JSON.stringify({
          ...processedData,
          id: Date.now().toString(),
          sheets_synced: sheetsSuccess
        })
      };
    }

    return {
      statusCode: 405,
      headers,
      body: JSON.stringify({ error: 'Method not allowed' })
    };

  } catch (error) {
    console.error('Function error:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: 'Internal server error' })
    };
  }
};
